src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,241 @@
+          }
+          continue;
+        }
+        if (ref_val_vector) {
+          if (t_comparator((*ref_val_vector)[r.chunk_offset], t_value)) {
+            pos_list->push_back(r);
+          }
+          continue;
+        }
+        DebugAssert(false, "Unexpected Column Type when dereferencing ReferenceColumn");
+      }
+      Chunk reference_chunk;
+
+      // create reference columns
+      for (ColumnID i{0}; i < ref_table->col_count(); ++i) {
+        auto reference_column = std::make_shared<ReferenceColumn>(ref_table, i, pos_list);
+        reference_chunk.add_column(reference_column);
+      }
+
+      output_table->emplace_chunk(std::move(reference_chunk));
+
+      return output_table;
+    }
+  }
+
+  auto ValueID_comparator = make_comparator<ValueID>(scan_type);
+  // search through every chunk of the input table
+  // store finds in pos_list
+  for (auto chunk_id = ChunkID(0); chunk_id < table->chunk_count(); ++chunk_id) {
+    const auto& chunk = table->get_chunk(chunk_id);
+    // get the column that will be searched
+    auto column = chunk.get_column(column_id);
+
+    // Columns can be of types ValueColumn or DictColumn
+    auto value_column = std::dynamic_pointer_cast<const ValueColumn<T>>(column);
+    if (value_column) {
+      // since ValueColumns are not sorted, just search through them linearly
+      const auto& values = value_column->values();
+      for (auto chunk_offset = ChunkOffset(0); chunk_offset < values.size(); ++chunk_offset) {
+        if (t_comparator(values[chunk_offset], t_value)) {
+          pos_list->push_back(RowID{chunk_id, chunk_offset});
+        }
+      }
+      continue;
+    }
+
+    auto dict_column = std::dynamic_pointer_cast<const DictionaryColumn<T>>(column);
+    DebugAssert(dict_column,
+                "Unknown Column Type in Table Scan");  // make sure the above is the case; only in debug mode
+
+    const auto dictionary = dict_column->dictionary();
+
+    // depending on the operator we need to select a proper ValueID
+    // which accurately represents the Value we are searching for
+    // (keep in mind it can be the case that the value we are searching for is not contained)
+
+    bool value_contained = std::binary_search(dictionary->cbegin(), dictionary->cend(), t_value);
+
+    if (!value_contained) {
+      if (scan_type == ScanType::OpEquals) {
+        // no value can be equal
+        continue;
+      }
+      if (scan_type == ScanType::OpNotEquals) {
+        add_all_values(*pos_list, chunk_id, dict_column->size());
+        continue;
+      }
+    }
+
+    ValueID comp_value;
+    switch (scan_type) {
+      case ScanType::OpEquals:
+      case ScanType::OpNotEquals:
+      case ScanType::OpLessThan:
+      case ScanType::OpGreaterThanEquals:
+        comp_value = dict_column->lower_bound(t_value);
+        break;
+      case ScanType::OpLessThanEquals:
+      case ScanType::OpGreaterThan:
+        comp_value = dict_column->upper_bound(t_value);
```

Comment:
I think using upper_bound to find the comp_value might be a logical problem (unless I'm missing some further transformation here): It *never* returns the position of t_value in the dictionary, but the position of the next-bigger element (if any). So I guess you're comparing the wrong value later on.
http://en.cppreference.com/w/cpp/algorithm/upper_bound